src/backend/renderer/multigpu/mod.rs

```diff
@@ -360,53 +360,38 @@ impl<A: GraphicsApi> GpuManager<A> {
                 if let Some(data) = states.data_map.get::<RefCell<RendererSurfaceState>>() {
                     let mut data_ref = data.borrow_mut();
                     let data = &mut *data_ref;
-                    let attributes = states.cached_state.current::<SurfaceAttributes>();
-                    // Import a new buffer if available
-                    let surface_size = data
-                        .buffer_dimensions
-                        .map(|d| d.to_logical(data.buffer_scale, data.buffer_transform));
-                    if let Some(buffer) = data.buffer.as_ref() {
-                        let surface_size = surface_size.unwrap();
-                        let buffer_damage = attributes
-                            .damage
-                            .iter()
-                            .flat_map(|dmg| {
-                                match dmg {
-                                    Damage::Buffer(rect) => *rect,
-                                    Damage::Surface(rect) => rect.to_buffer(
-                                        attributes.buffer_scale,
-                                        attributes.buffer_transform.into(),
-                                        &surface_size,
-                                    ),
-                                }
-                                .intersection(Rectangle::from_loc_and_size(
-                                    (0, 0),
-                                    data.buffer_dimensions.unwrap(),
-                                ))
-                            })
-                            .fold(Vec::<Rectangle<i32, BufferCoords>>::new(), |damage, mut rect| {
-                                // replace with drain_filter, when that becomes stable to reuse the original Vec's memory
-                                let (overlapping, mut new_damage): (Vec<_>, Vec<_>) =
-                                    damage.into_iter().partition(|other| other.overlaps(rect));
-
-                                for overlap in overlapping {
-                                    rect = rect.merge(overlap);
-                                }
-                                new_damage.push(rect);
-                                new_damage
-                            });
-
-                        if let Err(err) =
-                            self.early_import_buffer(source, target, buffer, states, &*buffer_damage)
-                        {
-                            result = Err(err);
+                    if data.textures.is_empty() {
+                        // Import a new buffer if available
+                        if let Some(buffer) = data.buffer.as_ref() {
+                            let buffer_damage = data.damage.iter().take(1).flatten().cloned().fold(
```

Comment:
What's the reason for using `take` here? The entire fold for damage will only be run once since `take(1)` will only yield one element?

Reply:
Yes correct. The early_import cannot know, if we have rendered the last commit. So it may very well be, that we require more damage (possibly also depending on the output/space it is rendered on), but we try to optimistically upload parts of the texture as an optimization. So doing a very defensive import here of everything we can upload kinda defeats the purpose.

If it ends up being to few parts copied, the multigpu-code will import the rest (see call to `import_missing`), but most of the time we render in lock-step with the application, given that hardware accelerated apps usually wait for frame-events. In these cases this should be sufficient, so we really want only one frame of damage.

Reply:
I see, might be worth documenting that behavior as those semantics aren't really described there.

Reply:
Well yeah, the whole renderer is probably lacking in terms of docs. but sure. I will add this bit, better to start somewhere I suppose.

Reply:
done